starknet-core/src/types/messaging.rs

```diff
@@ -0,0 +1,216 @@
+use alloc::vec::Vec;
+
+use num_bigint::BigUint;
+use sha3::{Digest, Keccak256};
+use starknet_ff::FieldElement;
+
+use super::{eth_address::EthAddress, L1HandlerTransaction, MsgToL1};
+use crate::utils::biguint_to_felts;
+
+pub struct MsgToL2 {
+    from_address: EthAddress,
+    to_address: FieldElement,
+    nonce: FieldElement,
+    selector: FieldElement,
+    payload: Vec<FieldElement>,
+}
+
+#[allow(clippy::vec_init_then_push)]
+impl MsgToL2 {
+    /// Returns the hash as a vector a FieldElements
+    /// representing a serialized u256: [low, high].
+    pub fn hash_as_felts(&self) -> Vec<FieldElement> {
+        // TODO: is it safe to unwrap as keccak returns a valid u256?
+        biguint_to_felts(self.hash()).unwrap()
+    }
+
+    /// Computes MsgToL2 hash.
+    pub fn hash(&self) -> BigUint {
+        let mut buf: Vec<u8> = Vec::new();
+
+        // As EthAddress is only 20 bytes, we do want the 32 bytes
+        // to compute the hash as done in solidity `uint256(uint160(fromAddress))`.
+        let from_felt: FieldElement = self.from_address.clone().into();
+
+        buf.extend(from_felt.to_bytes_be());
+        buf.extend(self.to_address.to_bytes_be());
+        buf.extend(self.nonce.to_bytes_be());
+        buf.extend(self.selector.to_bytes_be());
+        buf.extend(FieldElement::from(self.payload.len()).to_bytes_be());
+        for p in &self.payload {
+            buf.extend(p.to_bytes_be());
+        }
+
+        let mut hasher = Keccak256::new();
+        hasher.update(buf);
+        let hash = hasher.finalize();
+        BigUint::from_bytes_be(hash.as_slice())
+    }
+}
+
+impl L1HandlerTransaction {
+    /// Parses and returns the `MsgToL2` from
+    /// the transaction's content.
+    pub fn parse_msg_to_l2(&self) -> MsgToL2 {
+        // TODO: is that necessary? As the sequencer
+        // itself is the one firing this kind of transaction?
+        assert!(!self.calldata.is_empty());
```

Comment:
This `assert!` is completely unacceptable. You're assuming that whatever `self` instances always come from a trusted source, which is not true. As a library, you cannot decide this for your users. Only _invariants_ can be checked with panic in a library.

Reply:
got it, thank you for pointing this out.